File: web/src/cost.js
// Token formatting + cost DISPLAY (W2.5, then the per-model pricing fix). One home so the
// board, the drawer and Stats all say the same number — spend must be visible where approval
// decisions happen, not only on an aggregate page.
//
// THIS FILE NO LONGER COMPUTES A DOLLAR FIGURE. It used to (see the two repair stories below,
// kept as history): a flat `costOf` priced every attempt at one hardcoded Anthropic rate
// ($3/1K fresh, $0.3/1K cache-read), which was wrong for any Codex/OpenAI attempt the moment
// `core/pricing.py` gained per-model OpenAI rows — a `gpt-5.3-codex` attempt ($1.75/$14
// published) rendered on the board as if it had been billed at Sonnet's $3/$15. Fixing that in
// JS a third time (a second price table, or a defaulted-fallback rate baked into a `??`) would
// just move the drift somewhere else, so the fix moves pricing server-side: `core/cost.py`'s
// `attempt_cost` /
// `attempts_cost` price every attempt at ITS OWN recorded model via `core/pricing.py`, and the
// API sends the result as `cost_usd` (+ `cost_model`, naming what priced it) on every
// attempt/task/metrics payload. `taskCost` and `lifetimeCost` below now only READ those fields
// — no rate, no per-bucket arithmetic, nothing to drift a second time. The Python side of this
// split is pinned in `tests/test_pricing_usd.py`; the API wiring in `tests/test_api.py`.
//
// Cost HISTORY, kept because the failure modes it documents are still the reason the buckets
// below are named rather than positional, even though the arithmetic that used them is gone:
// the burn a card shows and the cost it is priced at must have the SAME basis — they didn't
// (the Token Usage tile once read "169.87M · est. $73.58", pricing 6M cache-creation tokens
// the count never showed), and a single-argument `estimateCost(tokens, cacheRead = 0)` let a
// one-argument call silently price a TOTAL burn at the fresh rate (how a Stats tile once
// claimed $29.98 per merged PR).

export function fmtTokens(n) {
  if (n == null) return "—";
  if (n < 1000) return `${n}`;
  if (n < 1000000) return `${(n / 1000).toFixed(1)}k`;
  return `${(n / 1000000).toFixed(2)}M`;
}

/**
 * Every token bucket, summed — the number a burn meter shows.
 *
 * Cache-read is 90%+ of real spend (C1): summing only tokens_used under-reported a 33M-token
 * task as "121.5k tok". Cache-CREATION was the bucket still missing, so the burn a surface
 * displayed and the cost it priced had different bases (the Token Usage tile read
 * "169.87M · est. $73.58" — a price for 6M tokens the count never showed).
 *
 * Named buckets: they cannot be transposed or silently dropped.
 */
export function totalBurn(buckets) {
  const { used = 0, creation = 0, read = 0 } = buckets || {};
  return (used || 0) + (creation || 0) + (read || 0);
}

/** Format a dollar figure — or say nothing, which beats saying a wrong number. */
export function fmtCost(dollars) {
  if (dollars == null || !Number.isFinite(dollars) || dollars === 0) return "—";
  if (dollars < 0.01) return "<$0.01";
  return `$${dollars.toFixed(2)}`;
}

/**
 * One task's token burn across the same nine buckets the API's `cost_usd` prices — so a
 * surface showing both cannot show a price for tokens its own count never included.
 */
export function taskBurn(task) {
  if (!task) return 0;
  return (
    totalBurn({ used: task.total_tokens, creation: task.total_cache_creation, read: task.total_cache_read })
    + totalBurn({
      used: task.total_review_tokens,
      creation: task.total_review_cache_creation,
      read: task.total_review_cache_read,
    })
    + totalBurn({
      used: task.total_aux_tokens,
      creation: task.total_aux_cache_creation,
      read: task.total_aux_cache_read,
    })
  );
}

/**
 * One task's cost — read straight off the API. Priced server-side by `core/cost.py`'s
 * `attempt_cost`/`attempts_cost`, each attempt at its OWN recorded model
 * (`core/pricing.py`'s per-model table), summed across coder + reviewer + aux. `task.cost_model`
 * names what priced it: a model id, `"mixed"` when the task's attempts used more than one, or
 * `pricing.FALLBACK_PRICE_NAME` when nothing priced could be resolved — never a rate computed
 * here. `0` for a task with no attempts yet or no cost field (an older payload shape), same as
 * before this file stopped computing.
 */
export function taskCost(task) {
  return task?.cost_usd ?? 0;
}

/**
 * The lifetime cost, from `/api/metrics`'s `cost_usd_total` — the SINGLE source both the
 * "Cost / merged PR" tile and the "Token Usage" tile read. `null` when the install has no
 * attempts yet (distinct from "attempts spent $0"), so both tiles say "—" together rather than
 * disagreeing.
 */
export function lifetimeCost(metrics) {
  return metrics?.cost_usd_total ?? null;
}

/**
 * Only `api_key` (BYO) mode pays Anthropic per token for real; `subscription`
 * (the default, and any absent/unrecognized value — OAuth is the fallback,
 * never assume real dollars) pays a flat fee, so a dollar figure there is an
 * API-rate ESTIMATE, not money that changed hands. The single home of this
 * rule (SCRUM re-home) — every surface that must choose tokens vs. dollars
 * reads this instead of re-deriving the `=== "api_key"` check.
 */
export function pricingIsReal(authMode) {
  return authMode === "api_key";
}

/**
 * Lifetime tokens from `/api/metrics`'s `tokens_total` — the token-basis
 * sibling of `lifetimeCost`, reading the SAME nine buckets `cost_usd_total`
 * prices. `null` on an older payload that lacks the key (distinct from "0
 * tokens spent"), so a subscription-mode tile can say "no token data yet"
 * instead of a false zero.
 */
export function lifetimeTokens(metrics) {
  return metrics?.tokens_total ?? null;
}

File: src/no_human/core/cost.py
"""One attempt's USD cost, and one task's — the ONLY place either is computed.

The bug this file exists to fix: ``web/src/cost.js`` priced every attempt at
one flat Anthropic rate ($3/1K fresh, $0.3/1K cache-read), which was simply
wrong for a Codex/OpenAI attempt once ``core/pricing.py`` gained per-model
OpenAI rows (``gpt-5.3-codex`` is $1.75/$14, not $3/$15) — the board was
pricing an 8x cheaper model as if it were Sonnet, silently, for every Codex
task on the page. Fixing it in JS a third time (a second price table, or a
``?? 3.0`` default) would just move the drift somewhere else; the API is now
the one place a USD figure is computed, and the board only formats what it is
sent. See ``web/src/cost.js``'s header for the JS side of that split.

``attempt_cost`` prices ONE attempt row: it walks every role in
``db.USAGE_ROLES``, looks up that role's model from the ``attempts.models``
JSON column (keyed on role NAME — "coder", "reviewer", ... — not the column
prefix), and sums ``pricing.usd_cost`` per role at that role's own rate. A
row where two roles used different models (e.g. a Codex coder reviewed by
Claude) is real and common, so the per-attempt label is the shared model when
every priced role agrees, and ``"mixed"`` when they do not — collapsing to
one of them would misreport which model actually spent which dollar.

DEGENERATE INPUT, all handled without raising, because every caller here is a
read path the board renders directly and a 500 is worse than an honest
number:

  * ``models`` is ``NULL``, ``'{}'``, or not valid JSON (11 of this install's
    684 attempt rows recorded before the column existed) — every role prices
    at ``pricing.FALLBACK_PRICE_NAME``, never at 0.0.
  * A role's token columns are all ``NULL``/0 — that role contributes $0 and
    its (unpriced) model does not count toward the "mixed" decision, so an
    attempt with real spend in only one role is never mislabeled "mixed" by
    an idle one.
  * The attempt list passed to ``attempts_cost`` is ``None``/empty — returns
    ``(None, None)``, matching every sibling ``total_*`` aggregate in
    ``api/models.py`` (``_aux_totals``): "no attempts" and "attempts that
    spent nothing" are different facts and must not both render as 0.
  * ``ledger_rows_as_attempts``: a row's ``model`` is ``None``, or an id
    ``pricing`` does not recognise — prices at ``FALLBACK_PRICE_NAME``, same
    as an attempt row would. A row's ``site`` does not match any registered
    ``AUX_USAGE_TIERS`` prefix (no such row exists today) — priced as
    ``"utility_"`` rather than dropped or raised, because a wrong-but-priced
    dollar is a smaller error than a silently missing one. ``rows`` is
    ``None``/empty — returns ``[]``, so ``attempts_cost(attempts + [])``
    behaves exactly as it does today with no ledger rows at all.
"""

from __future__ import annotations

import json
from typing import Any, Mapping

from .db import AUX_USAGE_TIERS, ORPHANED_SITE_PREFIX, USAGE_ROLES, usage_columns_for
from .pricing import FALLBACK_PRICE_NAME, input_price_usd_per_mtok, usd_cost


def _parse_models(raw: Any) -> dict[str, Any]:
    """``attempts.models`` as a role-name -> model-id dict, or ``{}``.

    Mirrors how it is written: ``Orchestrator._active_models()`` builds a
    plain dict and ``Store.update_attempt`` JSON-encodes it on the way in, so
    a healthy row holds a JSON object string. Anything else (NULL, ``''``,
    ``'{}'``, a non-JSON string, or JSON that decoded to something other than
    an object) is a row this pricer must still price, not reject — it prices
    at the fallback instead, exactly like an unpriced model id does.
    """
    if isinstance(raw, dict):
        return raw
    if isinstance(raw, str) and raw.strip():
        try:
            parsed = json.loads(raw)
        except (TypeError, ValueError):
            return {}
        if isinstance(parsed, dict):
            return parsed
    return {}


def attempt_cost(row: Mapping[str, Any] | None) -> tuple[float, str]:
    """One attempt's USD cost and the model label it was priced at.

    Sums ``pricing.usd_cost`` per role in ``db.USAGE_ROLES``, each at its own
    recorded model. Returns ``(dollars, label)`` — ``label`` is the one model
    every role with nonzero spend agreed on, ``pricing.FALLBACK_PRICE_NAME``
    when the row spent nothing chargeable, or ``"mixed"`` when priced roles
    disagree. Never raises: see the module docstring's DEGENERATE INPUT list.
    """
    if not row:
        return 0.0, FALLBACK_PRICE_NAME
    models = _parse_models(row.get("models"))
    total = 0.0
    labels: set[str] = set()
    for prefix, role in USAGE_ROLES.items():
        tokens_used_col, cache_read_col, cache_creation_col = usage_columns_for(prefix)
        output_col = "output_tokens" if prefix == "" else f"{prefix}output_tokens"
        tokens_used = row.get(tokens_used_col) or 0
        cache_read = row.get(cache_read_col) or 0
        cache_creation = row.get(cache_creation_col) or 0
        if not (tokens_used or cache_read or cache_creation):
            # This role spent nothing on this attempt — no dollars, and its
            # (possibly unpriced) model must not affect the mixed/agree call
            # for roles that actually spent something.
            continue
        model = models.get(role)
        dollars = usd_cost(
            tokens_used=tokens_used,
            cache_read_tokens=cache_read,
            cache_creation_tokens=cache_creation,
            output_tokens=row.get(output_col),
            model=model,
        )
        total += dollars
        _, label = input_price_usd_per_mtok(model)
        labels.add(label)
    if not labels:
        return 0.0, FALLBACK_PRICE_NAME
    if len(labels) == 1:
        return round(total, 6), next(iter(labels))
    return round(total, 6), "mixed"


def attempts_cost(rows: list[Mapping[str, Any]] | None) -> tuple[float | None, str | None]:
    """The USD cost of a list of attempts, summed, and its shared model label.

    ``(None, None)`` when ``rows`` is ``None``/empty — "this task has no
    attempts yet" — never ``(0.0, ...)``, which would say "this task's
    attempts cost nothing" and is not the same fact. Once there is at least
    one attempt, the total is always a number (0.0 is legitimate there: it
    means the attempts recorded really did spend zero).
    """
    if not rows:
        return None, None
    total = 0.0
    labels: set[str] = set()
    for row in rows:
        dollars, label = attempt_cost(row)
        total += dollars
        if dollars:
            labels.add(label)
    total = round(total, 6)
    if not labels:
        return total, FALLBACK_PRICE_NAME
    if len(labels) == 1:
        return total, next(iter(labels))
    return total, "mixed"


def ledger_rows_as_attempts(rows: list[Mapping[str, Any]] | None) -> list[dict[str, Any]]:
    """Reshape OWNED ``unattributed_usage`` rows (``Store.
    usage_ledger_rows_by_task`` / ``task_usage_ledger_rows``, or a
    whole-ledger grouping) into attempt-shaped dicts, so ``attempt_cost`` /
    ``attempts_cost`` price them at their own recorded model instead of a
    second, drifting pricer being written for the ledger.

    Each input row is one ``(site, model)`` group carrying
    ``tokens_used``/``cache_read_tokens``/``cache_creation_tokens``. The role
    is derived from ``site``: ``orphaned_<tier>usage`` strips
    ``ORPHANED_SITE_PREFIX`` and the trailing ``"usage"`` to recover
    ``<tier>`` (e.g. ``"orphaned_plan_usage"`` -> ``"plan_"``), validated
    against ``AUX_USAGE_TIERS`` — the ledger writer today writes nothing
    else. An unrecognised ``site`` falls back to ``"utility_"`` (see the
    module docstring's DEGENERATE INPUT list).

    No ``{tier}output_tokens`` key is emitted: ``unattributed_usage`` has no
    such column (it is a coder/reviewer SDK-only split), so the output
    premium is legitimately absent here, not a bug — same note
    ``orchestrator.py``'s aux-flush carries — and ``attempt_cost`` already
    treats a missing output column as zero extra, exactly as it does for any
    attempt row recorded before that column existed.

    The output is ordinary attempt-shaped mappings with only one role's
    columns populated, so ``attempts_cost`` prices each row at its own model
    and applies the identical "shared model, else mixed" label rule used for
    real attempts — one pricer, one mixed rule, for both tables.
    """
    out: list[dict[str, Any]] = []
    for row in rows or []:
        site = str(row.get("site") or "")
        tier: str | None = None
        if site.startswith(ORPHANED_SITE_PREFIX) and site.endswith("usage"):
            candidate = site[len(ORPHANED_SITE_PREFIX):-len("usage")]
            if candidate in AUX_USAGE_TIERS:
                tier = candidate
        if tier is None:
            tier = "utility_"
        role = USAGE_ROLES[tier]
        tokens_col, read_col, creation_col = usage_columns_for(tier)
        out.append({
            "models": {role: row.get("model")},
            tokens_col: row.get("tokens_used") or 0,
            read_col: row.get("cache_read_tokens") or 0,
            creation_col: row.get("cache_creation_tokens") or 0,
        })
    return out
